source/bdDetect.py

```diff
@@ -238,121 +237,119 @@ class Detector(object):
-							continue
-						if braille.handler.setDisplayByName(driver, detected=match):
-							return
-				if detectBluetooth:
-					if self._stopEvent.isSet():
-						continue
-					with self._btDevsLock:
-						if self._btDevs is None:
-							btDevs = list(getDriversForPossibleBluetoothDevices())
-							# Cache Bluetooth devices for next time.
-							btDevsCache = []
-						else:
-							btDevs = self._btDevs
-							btDevsCache = btDevs
-					for driver, match in btDevs:
-						if self._stopEvent.isSet() or (self._limitToDevices and driver not in self._limitToDevices):
-							continue
-						if btDevsCache is not btDevs:
-							btDevsCache.append((driver, match))
-						if braille.handler.setDisplayByName(driver, detected=match):
-							return
-					if self._stopEvent.isSet():
-						continue
-					if btDevsCache is not btDevs:
-						with self._btDevsLock:
-							self._btDevs = btDevsCache
+		if self._queuedFuture:
+			# This will cancel a queued scan (i.e. not the currently running scan, if any)
+			# If this future belongs to a scan that is currently running or finished, this does nothing.
+			self._queuedFuture.cancel()
+
+	# C901 '_bgScan' is too complex
+	# Note: when working on _bgScan, look for opportunities to simplify
+	# and move logic out into smaller helper functions.
+	def _bgScan(  # noqa: C901
+			self,
+			detectUsb: bool,
+			detectBluetooth: bool,
+			limitToDevices: typing.Optional[typing.List[str]]
+	):
+		"""Performs the actual background scan.
+		this function should be run on a background thread.
+		@param usb: Whether USB devices should be detected for this particular scan.
+		@param bluetooth: Whether Bluetooth devices should be detected for this particular scan.
+		@param limitToDevices: Drivers to which detection should be limited for this scan.
+			C{None} if no driver filtering should occur.
+		"""
+		# Clear the stop event before a scan is started.
+		# Since a scan can take some time to complete, another thread can set the stop event to cancel it.
+		self._stopEvent.clear()
+		if detectUsb:
+			if self._stopEvent.isSet():
+				return
+			for driver, match in getDriversForConnectedUsbDevices():
+				if self._stopEvent.isSet():
+					return
+				if (limitToDevices and driver not in limitToDevices):
+					continue
+				if braille.handler.setDisplayByName(driver, detected=match):
+					return
+		if detectBluetooth:
+			if self._stopEvent.isSet():
+				return
+			with self._btDevsLock:
+				if self._btDevs is None:
+					btDevs = list(getDriversForPossibleBluetoothDevices())
+					# Cache Bluetooth devices for next time.
+					btDevsCache = []
+				else:
+					btDevs = self._btDevs
+					btDevsCache = btDevs
+			for driver, match in btDevs:
+				if self._stopEvent.isSet():
+					return
+				if (self._limitToDevices and driver not in limitToDevices):
+					continue
+				if btDevsCache is not btDevs:
+					btDevsCache.append((driver, match))
+				if braille.handler.setDisplayByName(driver, detected=match):

```

Comment:
Not a new problem, but braille code would be easier to reason about if the braille display could only change at a deterministic time in the main thread loop. Perhaps this should instead queue a braille driver change to happen in the next core loop.

Reply:
I do like that suggestion!

Reply:
This is an interesting point, and as per the comment by @dkager https://github.com/nvaccess/nvda/pull/14147#issuecomment-1276429319 it is probably best to address this issue afterwards.

>how ought the braille detection thread to know that an attempt to connect to a display was succesful?

I haven't looked into the feasibility, but ideally there would be a separation of "setting the active display" and "attempting to connect to the display". If that were true, core would have a display driver instance that is already connected, ready to "install" as the active display driver at the appropriate change-over point. It would continue to use the old display up until that point.

Reply:
I think I'd prefer doing this in a follow up indeed.